### PR TITLE
[10.0][IMP] account_credit_control : Improve usability

### PR DESCRIPTION
**Pull request by rousseldenis**

### CODE REVIEW

**Comment by rousseldenis** [comment]
@gva-acsone @chrisdec @sbejaoui @sbidoul

**Comment by pedrobaeza** [comment]
We did some usability improvement in v11 in #695. You should check it for seeing matching things and align all versions.

**Comment by rousseldenis** [comment]
@pedrobaeza Yes, your changes are complementary, so integrate them should not be so complex.

I will note this in my todo for the sprint.

File: account_credit_control/models/account_invoice.py [hunk 1]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+        for invoice in invoices:
+            invoice_moves = moves.filtered(
+                lambda m: m.invoice_id.id == invoice.id).with_context(
+                from_parent_object=True)
+            for invoice_move in invoice_moves:
+                invoice.credit_control_notes =\
+                    invoice_move.credit_control_notes
+                break
+
+    @api.model
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py]:
```suggestion
    @api.multi
```

File: account_credit_control/models/account_invoice.py [hunk 2]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+    credit_control_date = fields.Date(
+        compute='_compute_credit_control_date',
+        inverse='_inverse_credit_control_date',
+        string='Credit Control Ignore Before',
+        readonly=True,
+        states={'open': [('readonly', False)]}
+    )
+
+    @api.multi
+    def _inverse_credit_control_notes(self):
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py:48]:
```suggestion:+16
    def _inverse_credit_control_notes(self):
        for invoice in self:
            invoice._get_payable_receivable_move_line().credit_control_notes = invoice.credit_control_notes
```

File: account_credit_control/models/account_invoice.py [hunk 3]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+            invoice_moves = moves.filtered(
+                lambda m: m.invoice_id == invoice.id)
+
+            for invoice_move in invoice_moves:
+                invoice_move.credit_control_notes =\
+                    invoice.credit_control_notes
+                break
+
+    @api.multi
+    def _compute_credit_control_notes(self):
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py:59]:
```suggestion:+11
    def _compute_credit_control_notes(self):
        for invoice in self:
            invoice.credit_control_notes = invoice._get_payable_receivable_move_line().credit_control_notes
```

File: account_credit_control/models/account_invoice.py [hunk 4]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+            invoice_moves = moves.filtered(
+                lambda m: m.invoice_id.id == invoice.id).with_context(
+                from_parent_object=True)
+            for invoice_move in invoice_moves:
+                invoice.credit_control_notes =\
+                    invoice_move.credit_control_notes
+                break
+
+    @api.model
+    def _get_all_related_move_line(self):
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py]: No need for this one according to other changes.
Replace it by:
```suggestion:+6
    def_get_payable_receivable_move_line(self):
        self.ensure_one()
        return self.move_id.line_ids.filtered(lambda l: l.account_id.internal_type in ['payable', 'receivable'])
```

File: account_credit_control/models/account_invoice.py [hunk 5]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+        # Search for the payable or receivable account move line
+        # (1 for each invoice)
+        return self.env['account.move.line'].search([
+            ('account_id.internal_type', 'in', ['payable', 'receivable']),
+            ('invoice_id', 'in', self.ids)
+        ])
+
+    @api.multi
+    def _inverse_credit_control_date(self):
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py:71]:
```suggestion:+9
    def _inverse_credit_control_date(self):
        for invoice in self:
            invoice._get_payable_receivable_move_line().credit_control_date = invoice.credit_control_date
```

File: account_credit_control/models/account_invoice.py [hunk 6]
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
...
+                lambda m: m.invoice_id.id == invoice.id).with_context(
+                from_parent_object=True)
+            for invoice_move in invoice_moves:
+                invoice_move.credit_control_date = invoice.credit_control_date
+                break
+
+    @api.multi
+    @api.depends('move_id.line_ids.credit_control_date',
+                 'move_id.line_ids.invoice_id')
+    def _compute_credit_control_date(self):
```
Comment (qgroulard) [inline account_credit_control/models/account_invoice.py:79]:
```suggestion:+10
    def _compute_credit_control_date(self):
        for invoice in self:
            invoice.credit_control_date = invoice._get_payable_receivable_move_line().credit_control_date
```

File: account_credit_control/models/res_partner.py [hunk 7]
```diff
@@ -51,6 +52,25 @@ class ResPartner(models.Model):
...
     )
+    credit_control_notes = fields.Char(
+        help="These are notes related to Credit Control for this partner",
+    )
+    credit_control_email = fields.Char(
+        help="This is the email used in place of standard email to manage"
+             " credit control"
+    )
+
+    _sql_constraints = [
```
Comment (qgroulard) [inline account_credit_control/models/res_partner.py]: This sql constraint doesn't make any sense in this module.

**Comment by rafaelbn** [comment]
Please @rousseldenis could you rebase? I can test this

**Comment by rousseldenis** [comment]
@rafaelbn We plan to work on this on OCA days

File: account_credit_control/models/account_invoice.py (hunk 2 above)
Comment (rousseldenis) [reply account_credit_control/models/account_invoice.py:48]: Done

File: account_credit_control/models/account_invoice.py (hunk 3 above)
Comment (rousseldenis) [reply account_credit_control/models/account_invoice.py:59]: Done

File: account_credit_control/models/account_invoice.py (hunk 5 above)
Comment (rousseldenis) [reply account_credit_control/models/account_invoice.py:71]: Done

**Comment by rousseldenis** [comment]
@qgroulard @rafaelbn This is done. Comments attended

File: account_credit_control/models/account_invoice.py [hunk 8]
```diff
@@ -30,6 +30,56 @@ class AccountInvoice(models.Model):
...
+        for invoice in self:
+            line = invoice._get_payable_receivable_move_line()
+            invoice.credit_control_notes = line.credit_control_notes
+
+    @api.multi
+    def _get_payable_receivable_move_line(self):
+        self.ensure_one()
+        return self.move_id.line_ids.filtered(
+            lambda l: l.account_id.internal_type in ['payable', 'receivable'])
+
```
Comment (houssine78) [inline account_credit_control/models/account_invoice.py:69]: from my understanding only receivable accounts are concerned by account_credit_control module... so why also filtering on payable ?

File: account_credit_control/models/res_partner.py [hunk 9]
```diff
@@ -51,6 +52,20 @@ class ResPartner(models.Model):
...
+    credit_control_email = fields.Char(
+        help="This is the email used in place of standard email to manage"
+             " credit control"
+    )
+
+    @api.model
+    def _get_default_credit_policy(self):
+        policy_obj = self.env['credit.control.policy']
+        default_policy = policy_obj.search([('default_on_partner', '=', True)])
+        if default_policy:
```
Comment (houssine78) [inline account_credit_control/models/res_partner.py:67]: should we take the do_nothing field into account.

File: account_credit_control/models/credit_control_policy.py [hunk 10]
```diff
@@ -29,6 +29,20 @@ class CreditControlPolicy(models.Model):
...
+    default_on_partner = fields.Boolean(
+        string="Default Value On Partner"
+    )
+
+    @api.constrains('default_on_partner')
+    @api.multi
+    def _constrains_default_on_partner(self):
+        count = self.search_count([
+            ('default_on_partner', '=', True)
+        ])
```
Comment (houssine78) [inline account_credit_control/models/credit_control_policy.py:41]: should we take the do_nothing field into account.

**Review by houssine78 — APPROVED** [review]
some small remarks but LGTM

File: account_credit_control/models/account_invoice.py (hunk 8 above)
Comment (rousseldenis) [reply account_credit_control/models/account_invoice.py:69]: I'm not quite used to accounting, @chrisdec a little help on this ?

Comment (rousseldenis) [reply account_credit_control/models/account_invoice.py:69]: You're right @houssine78. I'll change it

File: account_credit_control/models/res_partner.py (hunk 9 above)
Comment (rousseldenis) [reply account_credit_control/models/res_partner.py:67]: Hi @houssine78, as that field is on the policy, it does not avoid getting a default policy on the partner

File: account_credit_control/models/credit_control_policy.py (hunk 10 above)
Comment (rousseldenis) [reply account_credit_control/models/credit_control_policy.py:41]: Same as https://github.com/OCA/account-financial-tools/pull/819/files#r329928573

File: account_credit_control/models/res_partner.py (hunk 9 above)
Comment (houssine78) [reply account_credit_control/models/res_partner.py:67]: @rousseldenis my understanding of this field is to set it at True when the policy become obsolete. as it obsolete does it should be taken as default ? I mean people could forget to unset it on a partner. The idea is why it should continue to be returned as default if it's obsolete

Comment (qgroulard) [reply account_credit_control/models/res_partner.py:67]: @houssine78 The field 'default_on_partner' is on the policy not on the partner. Thus forgetting to unset it isn't really a concern imo.

Comment (houssine78) [reply account_credit_control/models/res_partner.py:67]: ok it's not blocking anyway. ok to move on as it is.

**Comment by qgroulard** [comment]
@rousseldenis

File: account_credit_control/models/account_move_line.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+class AccountMoveLine(models.Model):
+
+    _inherit = 'account.move.line'
+
+    credit_control_notes = fields.Char()
+    credit_control_date = fields.Date(
+        string='Credit Control Ignore Before',
+        help="If filled in, specify a pivot date to ignore moves before for "
+             "credit control"
```
Comment (sbidoul) [inline account_credit_control/models/account_move_line.py:15]: This help text is unclear to me.

**Review by sbidoul — COMMENTED** [review]
To facilitate review of this relatively big PR, it would be nice to have a functional summary of the changes, possibly with some screenshots.

**Comment by qgroulard** [comment]
Here is a list of changes:

- Add credit control notes (free text) on lines, move_lines, invoices and partners;
- Add a credit control date on invoices and move_lines to ignore Credit Control runs until this date;
- Add 'phone' channel;
- Therefore rename to_be_send -> to_do, sent -> done and date_sent -> date_reminded;
- track_visibility="on_change" on credit control line state;
- Display level_lines max level when grouped (groupby in tree view) instead of sum;
- Add a group by move_line_journals on credit control lines;
- Fallback to letter if channel is 'email' and partner has no email;
- Add config on policy to set the policy has default on partners;
- Dates format according to current user settings;
- Fix error message when trying to create a run at a previous date than last run;
- Add specific email for credit control on partner;

**Pull request closed without merging**